### PR TITLE
check that time for type inference in splatting arrays is not excessive

### LINKED ISSUE #17903: Array type inference takes exponential time
When concatenating a `Float64[]` and an `Any[]` that contains some `nothing` values, inferring the type of the resulting array takes time exponential in the length of the arrays. Example:

```
julia> for n in [100, 150, 200, 250, 300, 350, 400]
           x = zeros(Float64, n)
           y = [0.0, nothing]
           print("n=$n: ")
           @time result = [x..., y...]
       end
n=100:   0.711594 seconds (208.66 k allocations: 12.082 MB, 3.53% gc time)
n=150:   1.409423 seconds (88.71 k allocations: 4.886 MB)
n=200:   2.961411 seconds (100.57 k allocations: 5.549 MB)
n=250:   7.311390 seconds (126.25 k allocations: 6.930 MB, 0.07% gc time)
n=300:  16.698541 seconds (153.91 k allocations: 8.400 MB)
n=350:  25.183418 seconds (183.35 k allocations: 9.966 MB, 0.02% gc time)
n=400:  39.814280 seconds (214.75 k allocations: 11.587 MB)

julia> versioninfo()
Julia Version 0.4.6
Commit 2e358ce (2016-06-19 17:16 UTC)
Platform Info:
  System: Linux (x86_64-unknown-linux-gnu)
  CPU: Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

Workarounds: Changing `[x..., y...]` to `Any[x..., y...]` or `vcat(x, y)` does not exhibit this issue.

Background: In my case, the presence of `nothing` in `y` was a bug. It was a difficult bug to find though, because x had tens of thousands of items, so my program just hung with 100% CPU usage. When x and y are small, there is no noticeable problem, so it took me a while to figure out what was going on. It would be great if there was a way to set time limits on type inference, or at least get warnings if it's taking too long.

### PR DESCRIPTION
**Pull request by KristofferC**
Closes #17903

### CODE REVIEW

File: test/inference.jl [hunk 1]
```diff
@@ -524,3 +524,9 @@ f11015(a::AT11015) = g11015(Base.fieldtype(typeof(a), :f), true)
 g11015(::Type{Bool}, ::Bool) = 2.0
 @test Int <: Base.return_types(f11015, (AT11015,))[1]
 @test f11015(AT11015(true)) === 1
+
+# issue 17903
+n_17903 = 1000
+x_17903 = zeros(Float64, n)
+y_17903 = [0.0, nothing]
+@test (@elapsed [x_17903..., y_17903...]) < 30
```
Comment (yuyichao) [inline test/inference.jl]: This test is going to be problematic.....

Comment (KristofferC) [inline test/inference.jl]: I thought 30 was enough for it not to be problematic :P

Comment (KristofferC) [inline test/inference.jl]: Any better idea?

Comment (yuyichao) [inline test/inference.jl]: This should be a performance regression test.

Comment (KristofferC) [inline test/inference.jl]: Ok, makes sense.

File: test/inference.jl [hunk 2]
```diff
@@ -524,3 +524,9 @@ f11015(a::AT11015) = g11015(Base.fieldtype(typeof(a), :f), true)
 g11015(::Type{Bool}, ::Bool) = 2.0
 @test Int <: Base.return_types(f11015, (AT11015,))[1]
 @test f11015(AT11015(true)) === 1
+
+# issue 17903
+n_17903 = 1000
+x_17903 = zeros(Float64, n)
```
Comment (yuyichao) [inline test/inference.jl]: Also `n_17903`

**Comment by tkelman** [comment]
something to this effect should be added to nanosoldier benchmarks

**Pull request closed without merging**